src/i18n.ts

```diff
@@ -0,0 +1,248 @@
+  }
+
+  /** Returns a middleware to .use on the `Bot` instance. */
+  middleware(): Middleware<C & I18nContextFlavor> {
+    const fluent = this.fluent;
+    const { defaultLocale, localeNegotiator, useSession } = this.config;
+    return async function (
+      ctx: C & I18nContextFlavor,
+      next: NextFunction,
+    ): Promise<void> {
+      let translate: TranslateFunction;
+
+      ctx.i18n = {
+        fluent,
+        reNegotiateLocale: negotiateLocale,
+        useLocale,
+        getLocale: getNegotiatedLocale,
+        setLocale,
+      };
+      ctx.t = translateWrapper;
+      ctx.translate = translateWrapper;
+
+      await negotiateLocale();
+      await next();
+
+      // Also exports ctx object properties for accessing them directly from
+      // the translation source files.
+      function translateWrapper(
+        messageId: string,
+        context?: TranslationContext,
+      ): string {
+        return translate(messageId, {
+          first_name: ctx.from?.first_name ?? "",
+          ctx: JSON.stringify(makeContextObject(ctx)),
+          ...context,
+        });
+      }
+
+      async function getNegotiatedLocale(): Promise<string> {
+        return await localeNegotiator?.(ctx) ??
+          // deno-lint-ignore no-explicit-any
+          (await (useSession && (ctx as any).session))?.__language_code ??
+          ctx.from?.language_code ??
+          defaultLocale;
+      }
+
+      // Determining the locale to use for translations
+      async function negotiateLocale(): Promise<void> {
+        const negotiatedLocale = await getNegotiatedLocale();
+        useLocale(negotiatedLocale);
+      }
+
+      function useLocale(locale: LocaleId): void {
+        translate = fluent.withLocale(locale);
+      }
+
+      async function setLocale(locale: LocaleId): Promise<void> {
+        if (!useSession) {
+          throw new Error(
+"You are calling `ctx.i18n.setLocale()` without setting `useSession` \
+to `true` in the configuration. It doesn't make sense because you cannot set a \
+locale in session that way. You should either enable sessions or use \
+`ctx.i18n.useLocale()` instead.",
+          );
+        }
+
+        // deno-lint-ignore no-explicit-any
+        (await (ctx as any).session).__language_code = locale;
+        await negotiateLocale();
+      }
+    };
+  }
+}
+
+function makeContextObject(ctx: Context) {
+  const keys: Array<keyof Context> = [
+    "callbackQuery",
+    "channelPost",
+    "chatJoinRequest",
+    "chatMember",
+    "chat",
+    "chosenInlineResult",
+    "editedChannelPost",
+    "editedMessage",
+    "from",
+    "match",
+    "me",
+    "msg",
+    "message",
+    "inlineQuery",
+    "myChatMember",
+    "poll",
+    "pollAnswer",
+    "preCheckoutQuery",
+    "senderChat",
+    "shippingQuery",
+    "update",
+  ];
+
+  const obj: Record<string, unknown> = {};
+  for (const key of keys) obj[key] = ctx[key];
+  return obj;
+}
```

Comment:
Please correct me if I am wrong, or suggest me if there is a better way. But when I did `JSON.parse(ctx)`, it didn't contain any short-hand syntax such as `callbackQuery`, `chat`, `msg`, `from`, etc. But the point of having the `CTX` function in the translation context, is to simply access properties of the bot context.

So, if I just stringify the ctx, users have to write

```fluent
x = { CTX("update.message.from.username") }
```

where they can write,

```fluent
x = { CTX("from.username") }
```

to achieve the same.

I found it strange too. But, couldn't find any other way to stringify these.

Why stringify? I tried several ways to pass the bot context as a fluent variable. But failed to. In the end, I had to stringify it.

Reply:
Interesting stuff. I assumed it would be possible without turning every context object into a string. But given that we have to do that, this is fine. Feel free to resolve after reading.

Reply:
There has to be a way. I just failed to accomplish it that way. Maybe we can turn this also into an issue, if it is actually possible.